pkg/ring/tokens.go

```diff
@@ -0,0 +1,99 @@
+package ring
+
+import (
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"os"
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/prometheus/tsdb/fileutil"
+)
+
+const (
+	// TokensVersion1 is the version is a simple list of tokens.
+	TokensVersion1 = 1
+)
+
+// Tokens is a simple list of tokens.
+type Tokens []uint32
+
+// StoreToFile stores the tokens in the given directory.
+func (t Tokens) StoreToFile(tokenFilePath string) error {
+	if tokenFilePath == "" {
+		return errors.New("path is empty")
+	}
+
+	f, err := ioutil.TempFile(os.TempDir(), "tokens")
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		// If the file was not closed, then there must already be an error, hence ignore
+		// the error (if any) from f.Close(). If the file was already closed, then
+		// we would ignore the error in that case too.
+		f.Close()
+		// RemoveAll returns no error when tmp doesn't exist so it is safe to always run it.
+		if err := os.RemoveAll(f.Name()); err != nil {
+			level.Warn(util.Logger).Log("msg", "error deleting temporary file", "err", err)
+		}
+	}()
+
+	b, err := t.Marshal()
+	if err != nil {
+		return err
+	}
+	if _, err = f.Write(b); err != nil {
+		return err
+	}
+
+	if err := f.Close(); err != nil {
+		return err
+	}
+
+	// Block successfully written, make visible and remove old ones.
+	return fileutil.Replace(f.Name(), tokenFilePath)
```

Comment:
I think we should just use `os.Rename` here. `fileutil.Replace` has additional step: if target is a directory, it **removes** the directory, and then calls `os.Rename`, but I would argue that if target is indeed a directory (?), it's better to let writing of the token fail, keep the directory and report the error. Operator should handle this situation, cortex shouldn't just delete directories it doesn't like.

Reply:
It is a file and not a directory. And I see that `fileutil.Replace` takes care of syncing the changes, so I would prefer keeping it as it is. (And I realised that I copied the comment from Prometheus while I was hunting for this function without modifying it :) )

Reply:
My point was that `os.Rename` and `fileutil.Replace` differ in functionality only if target is a directory (which it isn't here), but you also point out sync on parent dir. I'm not convinced it's enough to justify dependency on fileutil package here.

Reply:
I don't have any strong opinion here, will change it to `os.Rename`.